Cluster contextualization

pejirout opened the pull request:
User may optionally provide additional information in the 'ucernvm'
section of the context file to use a master/slave cluster
contextualization.

Master sends its IP address to the central service (cernvm-online.cern.ch),
from where it is fetched by the slave machines. The slave machines
are waiting for the master to publish the data (blocking wait in the
bootloader).

Commit:
Cluster contextualization

User may optionally provide additional information in the 'ucernvm'
section of the context file to use a master/slave cluster
contextualization.

Master sends its IP address to the central service (cernvm-online.cern.ch),
from where it is fetched by the slave machines. The slave machines
are waiting for the master to publish the data (blocking wait in the
bootloader).

jblomer commented on scripts.d/06context:
```diff
@@ -324,3 +324,238 @@ if [ ! -b "${ROOT_DEV}" ]; then
+
+    # erase the first occurence of 'cluster_master=' from the grepped master line
+    master_field=$( echo "$master_field" | sed -e "s/$CLUSTER_MASTER_CONTEXT_FIELD=//1" 2>/dev/null )
+    if [ "x$master_field" = "xtrue" -o "x$master_field" = "xTrue" -o "x$master_field" = "x1" -o "x$master_field" = "xyes" ]; then
```
This can become a utility function in runtime-init.sh, like `IsOn`

jblomer commented on scripts.d/06context:
```diff
@@ -324,3 +324,238 @@ if [ ! -b "${ROOT_DEV}" ]; then
+    local CURL_CMD_TEMPLATE="curl -s -w \n\n%{http_code} -XGET"
+
+    # Get the 'master ready' status
+    local CURL_CMD="$CURL_CMD_TEMPLATE $SERVICE_URL/api/clusters/$CLUSTER_PIN/keys/$READY_STATUS_KEY_FIELD $CURL_HEADERS"
```
I think the API should have a version component, like `.../api/v1/clusters/...`

pejirout replied:
Ok, will do.

jblomer commented on scripts.d/06context:
```diff
@@ -324,3 +324,238 @@ if [ ! -b "${ROOT_DEV}" ]; then
+
+    local response=
+    local cnt=0
+    while [ $cnt -lt $MAX_TRIES ]; do
```
Making a curl request can become a function, I think it is the same code as further down.

pejirout replied:
It is, except in the second case we also need the get the response (not just the  status code). Because we cannot return two values from a function, then we would need to create yet another function to abstract it away, which I thought would be quite cumbersome (in comparison to the gained benefit).

Do you still want a special function for that?

jblomer replied:
Hm, perhaps code and ouput can be combined somehow.  For instance, first line status code, rest output.

jblomer commented on scripts.d/06context:
```diff
@@ -324,3 +324,238 @@ if [ ! -b "${ROOT_DEV}" ]; then
+    masterIp=$( echo "$response" | sed "s/$MASTER_IP_FIELD: //g" )
+    sed -i "s/$PLACECEHOLDER_REPLACEMENT/$masterIp/g" $CONTEXT_FILE 2>/dev/null
+    log_ok
+    log_start "Replaced the IP placeholder with: $masterIp"
```
This logline should additionally be added to /var/log/context.log

pejirout replied:
Ok.

jblomer commented on scripts.d/06context:
```diff
@@ -324,3 +324,238 @@ if [ ! -b "${ROOT_DEV}" ]; then
+    log_start "Master cluster contextualization: setting master keys"
+
+    # Extract the ip address of the master
+    INTERFACE_NAME=$( $BBOX ifconfig | head -n 1 | cut -d: -f1 | awk '{print $1}' )
```
Isn't there already an ifconfig symlink to busybox present?

pejirout replied:
It is, but I want to be sure I'm using the busybox ifconfig version. The system default ifconfig is producing a slightly different output format, which is a problem for us, because we need to parse the output. Here it should work fine, I just want to make that explicit.

jblomer replied:
OK

jblomer reviewed (changes requested):
Nice!  Besides the inline comments:

  - I think there is enough code to justify another snippet, e.g. 08clustercontext
  - The parsing of the [ucernvm-begin]..[ucernvm-end] block has been already done.  All the variables in this block are available as bash environment variables `_UCONTEXT_<VARIABLE NAME>`.  See for instance 14cvmfs.
  - I guess you didn't try this yet with an HTTPS URL, right?  I think we need to add a CA directory to the image that curl then picks up for HTTPS verification.

Commit:
Code polishing and restructuring

- REST API is now versioned
- curl request in slave contextualization is now a function
- Use already parsed variables from the ucernvm section
- Cluster contextualization part is now in a separate file
- Created IsOnMaster utility function

pejirout commented:
Trying to use curl with HTTPS in the bootloader phase is failing:
`Protocol https not supported or disabled in libcurl`

curl version:
`curl 7.48.0 (x86_64-pc-linux-gnu) libcurl/7.48.0 zlib/1.2.7
Protocols: http
Features: IPv6 Largefile libz UnixSockets`

Commit:
Added documentation for cluster contextualization
